llama.cpp

```diff
@@ -3228,7 +3228,31 @@ static struct ggml_cgraph * llm_build_llama(
         }
     }
 
+    int32_t * run_layer = batch.run_layers;
+    bool run_attn = false, run_mlp = false;
+    cur = inpL;
+
     for (int il = 0; il < n_layer; ++il) {
+        run_attn = run_mlp = true;
+        if (run_layer != NULL) {
+            if (*run_layer >= 0) {
+                run_attn = (*run_layer & 1) == 0;
+                run_mlp  = (*run_layer & 2) == 0;
+                run_layer++;
+            } else {
+                run_layer = NULL;
+            }
+        } else if (ggml_allocr_is_measure(lctx.alloc) && il == n_layer - 1) {
```

Comment:
Does anyone have any idea of why this stuff is needed? It actually doesn't even seem to work with 70B, in fact it fails right at the beginning:

```plaintext
perplexity: 23.34 seconds per pass - ETA 4 hours 14.77 minutes
[1]3.5298,[2]3.6875,[3]4.2038,[4]3.7584,
TEST A  0 + [] - len:   1, best:(? -1 @ 0.000), last took 88.19 sec
ggml_allocr_alloc: not enough space in the buffer (needed 58720256, largest block available 41943072)
GGML_ASSERT: ggml-alloc.c:148: !"not enough space in the buffer"
```

Something about skipping just part of the layer is making it need more memory, but I really don't understand what. (It also seemed to only happen when skipping part of the last layers on smaller models, but not in this case apparently.)

Unfortunately, I really don't understand what's going here and my workaround doesn't seem reliable.

Reply:
The `else if (ggml_allocr_is_measure(...) ..` branches should not be needed.

Not sure why it fails, but I think it's not impossible that the graph with skipped layers to occupy more memory than the full graph. Imagine that we have the full graph:

```
op:  A  M  A  M  A  M  A  M  A  M  A  M  ...
buf: a0 m0 a0 m0 a0 m0 a0 m0 a0 m0 a0 m0 ...
```

The results from the attention (A) are stored in the buffer `a0` and the results from the MLP (M) are stored in `m0`.
In this case we need total memory of `a0 + m0` because when computing the `i`th MLP for example, we need just the `a0` buffer from the previous attention.

Let's assume the buffer size `m0 > a0` and we skip all attention layers.
Then we would have:

```
op:  M  M  M  M  M  M  ...
buf: m0 m1 m0 m1 m0 m1 ...
```

We now need two MLP buffers to perform the computation `m0 + m1`, because for each MLP we need the results from the previous one and a separate buffer to store the results of the current MLP.
This is more memory than in the first case.

This is some intuition why the truncated graph might require more mem to evaluate, but not sure if this explains your observation.

Reply:
If you want to look into this, you can define `GGML_ALLOCATOR_DEBUG` in `ggml-alloc.c` to see what tensors are allocated during peak memory usage.

Reply:
Thanks! (And I actually did try debugging stuff, but it was failing the allocate the output tensor not one of the repeating ones so I wasn't able to figure out the problem that way)

GG's explanation helped (thanks again), so I was able to add a hack that mostly avoids the problem.